lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nfs
+
+import (
+	"net/url"
+	"os"
+	"path"
+
+	"github.com/vmware/vic/lib/portlayer/storage"
+	"github.com/vmware/vic/lib/portlayer/util"
+	"github.com/vmware/vic/pkg/trace"
+	"io/ioutil"
+)
+
+const (
+	// This is the same as our vmdk approach namespacing the data further on the filesystem
+	VolumesDir = "volumes"
+
+	// path that namespaces the metadata for a specific volume
+	metadataDir = "volumedata"
+
+	//Stock file permissions that are set, In the future we may pass these in.
+	filePermissions = 0755
+)
+
+type NFSv3VolumeStore struct {
+	//volume store name
+	Name string
+
+	//volume directory path, used to construct paths.
+	volumeStoreDir string
+
+	//handler for establishing connection to a target.
+	MH MountHandler
+
+	//nfs target for filesystem interaction
+	Target *url.URL
+
+	//Service selflink to volume store.
+	SelfLink *url.URL
+}
+
+func NewVolumeStore(op trace.Operation, storeName string, nfsTargetURL *url.URL, mount MountHandler) (*NFSv3VolumeStore, error) {
+	// XXX: Potential credential leak, maybe log less...
+	op.Infof("Creating datastore (%s) at target (%q)", storeName, nfsTargetURL.String())
+
+	target, err := mount.Mount(nfsTargetURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer mount.Unmount(target)
+
+	//we assume that nfsTargetURL.path already exists.
+	//make volumes directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, VolumesDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	//make metadata directory
```

Comment:
It is made here during the creation of a VolumeStore. Then it is used on line 104 in function `volMetadataDirPath(ID string)` which will construct the path to a particular volumes metadata. Then in that directory in the `writeMetadata` function we write out the individual keys in the info map as flat binary files.